### PR TITLE
[SYCL-MLIR][NFC] Refactor TypeSwitch instances

### PR DESCRIPTION
**Pull request by victor-eds**
1. Remove Default branches when not needed.
2. Merge equivalent branches.

Signed-off-by: Victor Perez <victor.perez@codeplay.com>

### CODE REVIEW

**Review by whitneywhtsang — APPROVED** [review]

File: mlir-sycl/lib/Dialect/IR/SYCLOpsDialect.cpp [hunk 1]
```diff
@@ -198,7 +198,7 @@ SYCLOpAsmInterface::getAlias(mlir::Type Type, llvm::raw_ostream &OS) const {
            << "_" << Ty.getNumElements() << "_";
         return AliasResult::FinalAlias;
       })
-      .Default([](auto) { return AliasResult::NoAlias; });
+      .Default(AliasResult::NoAlias);
```
[the remark this answers is not in the document]
Comment (etiotto) [reply mlir-sycl/lib/Dialect/IR/SYCLOpsDialect.cpp:205]: Ok that means returning ? Not very intuitive, the original code is clearer so I'd leave it in. 
But up to you, not feeling strongly about this.

File: polygeist/tools/cgeist/Lib/ValueCategory.cc [hunk 2]
```diff
@@ -109,9 +109,7 @@ void ValueCategory::store(mlir::OpBuilder &builder, mlir::Value toStore) const {
     // Ad-hoc extension of booleans
     auto ElementType =
         llvm::TypeSwitch<Type, Type>(val.getType())
-            .Case<LLVM::LLVMPointerType>(
-                [](auto Ty) -> Type { return Ty.getElementType(); })
-            .Case<MemRefType>(
+            .Case<MemRefType, LLVM::LLVMPointerType>(
                 [](auto Ty) -> Type { return Ty.getElementType(); })
             .Default([](auto) -> Type { llvm_unreachable("Unhandled type"); });
```
Comment (etiotto) [inline polygeist/tools/cgeist/Lib/ValueCategory.cc:114]: Remove the default just like in other places ?

**Review by etiotto — APPROVED** [review]
LGTM - left some comments inline.

File: polygeist/tools/cgeist/Lib/ValueCategory.cc (hunk 2 above)
Comment (victor-eds) [reply polygeist/tools/cgeist/Lib/ValueCategory.cc]: Missed that one. Fixed.

File: mlir-sycl/lib/Dialect/IR/SYCLTraits.cpp [hunk 3]
```diff
@@ -15,22 +15,14 @@
...
-      .Case<mlir::sycl::AccessorType>(GetDimension)
-      .Case<mlir::sycl::RangeType>(GetDimension)
-      .Case<mlir::sycl::IDType>(GetDimension)
-      .Case<mlir::sycl::ItemType>(GetDimension)
-      .Case<mlir::sycl::NdItemType>(GetDimension)
-      .Case<mlir::sycl::NdRangeType>(GetDimension)
-      .Case<mlir::sycl::GroupType>(GetDimension)
-      .Default(
-          [](auto) -> unsigned { llvm_unreachable("Invalid input type"); });
+      .Case<mlir::sycl::AccessorType, mlir::sycl::RangeType, mlir::sycl::IDType,
```
[the remark this answers is not in the document]
Comment (victor-eds) [reply mlir-sycl/lib/Dialect/IR/SYCLTraits.cpp]: Done

File: mlir-sycl/lib/Dialect/IR/SYCLOpsDialect.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (victor-eds) [reply mlir-sycl/lib/Dialect/IR/SYCLOpsDialect.cpp:205]: I see it as follows: `TypeSwitch` is just a structure holding an optional element which will be instantiated when one of the `Case`s matches. If no case matches, the `Default` value is the value `TypeSwitch` will hold.

**Pull request merged**